### PR TITLE
[Calling][Feature] Implementation of UFD diagnostics notification banners

### PR DESCRIPTION
**Pull request by lucianopa-msft**
## Purpose

* This implements handle of User Facing Diagnostics feature in the UISDK as notification banners in call screen following the specified design: https://www.figma.com/file/BOeihN2NJE7ROdyqbihCxi/Mobile-UI-Library?type=design&node-id=7588-30394&mode=design&t=Jp08jcHG0gYzeXF3-0
* 

## Does this introduce a breaking change?

- [ ] Yes
- [x] No

## Pull Request Type
What kind of change does this Pull Request introduce?

- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Documentation content changes
- [ ] Other... Please describe:

## Pull request checklist

This PR has considered:
- [x] Color Theming
- [x] iOS supported versions (all major versions greater than or equal current target deployment version)
- [x] VoiceOver and Keyboard Accessibility
- [x] Internationalization and Right to Left layouts
- [x] Size classes and window sizes (iPhone vs iPad, notched devices, multitasking, different window sizes, etc)
- [x] View Data Injection
- [x] Demo App UI/UX update

## How to Test

There a couple of diagnostics a some are hard to test because they indicate a underlying device issue, but a few ones to test may be:
Speaking While Muted
* Join a call
* Mute yourself 
* Start speaking 
* After some time we should see the banner

Network Quality
* Before join a call switch on [network link conditioner iOS feature](https://nshipster.com/network-link-conditioner/) to a bad network.
* Join a call
* Is likely to see a bad network banner 
* Also turn on/off internet should give reconnecting state.

## What to Check

| Before                                       | After                                      |
|----------------------------------------------|--------------------------------------------|
| Screenshot or description before this change | Screenshot or description with this change |

### CODE REVIEW

File: AzureCommunicationUI/sdk/AzureCommunicationUICalling/Sources/Presentation/SwiftUI/ViewComponents/Diagnostics/CallDiagnosticsViewModel.swift [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+            update(diagnosticModel: mediaDiagnostic)
+        }
+    }
+
+    private func update(diagnosticModel: NetworkDiagnosticModel) {
+        updateBottomToast(isBadState: diagnosticModel.value,
+                          viewModel: BottomToastDiagnosticViewModel(
+                                        localizationProvider: localizationProvider,
+                                        networkDiagnostic: diagnosticModel.diagnostic),
+                          where: { $0.networkDiagnostic == diagnosticModel.diagnostic })
```
[the remark this answers is not in the document]
Comment (JoshuaLai) [reply AzureCommunicationUI/sdk/AzureCommunicationUICalling/Sources/Presentation/SwiftUI/ViewComponents/Diagnostics/CallDiagnosticsViewModel.swift:48]: Cause it looks like your method below is already doing something similar to what I suggested above

File: AzureCommunicationUI/sdk/AzureCommunicationUICalling/Sources/Service/Calling/AzureCommunicationCalling/ACSDiagnosticQualityExtension.swift [hunk 2]
```diff
@@ -0,0 +1,23 @@
+//
+//  Copyright (c) Microsoft Corporation. All rights reserved.
+//  Licensed under the MIT License.
+//
+
+import AzureCommunicationCalling
+
+extension AzureCommunicationCalling.DiagnosticQuality {
```
Comment (JoshuaLai) [inline AzureCommunicationUI/sdk/AzureCommunicationUICalling/Sources/Service/Calling/AzureCommunicationCalling/ACSDiagnosticQualityExtension.swift:8]: Consider moving this under our `Utilities` folder where we have some other helpers / extension files

File: AzureCommunicationUI/sdk/AzureCommunicationUICalling/Sources/Service/Calling/AzureCommunicationCalling/CallingSDKEventsHandler.swift [hunk 3]
```diff
@@ -204,8 +223,152 @@ extension CallingSDKEventsHandler: CallDelegate,
         }
         dominantSpeakersSubject.send(speakers)
     }
+
     func call(_ call: Call, didChangeMuteState args: PropertyChangedEventArgs) {
         isLocalUserMutedSubject.send(call.isMuted)
     }
 
+    // MARK: NetworkDiagnosticsDelegate
```
Comment (JoshuaLai) [inline AzureCommunicationUI/sdk/AzureCommunicationUICalling/Sources/Service/Calling/AzureCommunicationCalling/CallingSDKEventsHandler.swift:231]: nitpick just watch out for / clean up our debug

**Comment by JoshuaLai** [comment]
Super excited to see UFD coming to the mobile composites. I mainly just have questions about the 
``` swift
typealias NetworkDiagnosticModel = CallDiagnosticModel<NetworkCallDiagnostic, Bool>
typealias MediaDiagnosticModel = CallDiagnosticModel<MediaCallDiagnostic, Bool>
```
Probably more so the `NetworkDiagnosticModel` using a Bool instead of the actual network state

File: AzureCommunicationUI/sdk/AzureCommunicationUICalling/Sources/Service/Calling/AzureCommunicationCalling/ACSDiagnosticQualityExtension.swift (hunk 2 above)
Comment (lucianopa-msft) [reply AzureCommunicationUI/sdk/AzureCommunicationUICalling/Sources/Service/Calling/AzureCommunicationCalling/ACSDiagnosticQualityExtension.swift:8]: But all those extensions that do the same `AzureCommunicationCalling` enum to a model enum conversion are in this same folder ...

File: AzureCommunicationUI/sdk/AzureCommunicationUICalling/Sources/Service/Calling/AzureCommunicationCalling/CallingSDKEventsHandler.swift (hunk 3 above)
Comment (lucianopa-msft) [reply AzureCommunicationUI/sdk/AzureCommunicationUICalling/Sources/Service/Calling/AzureCommunicationCalling/CallingSDKEventsHandler.swift:231]: Sorry, I don't understand this. Are you suggesting to remove the IDE `MARK:`?

Comment (lucianopa-msft) [reply AzureCommunicationUI/sdk/AzureCommunicationUICalling/Sources/Service/Calling/AzureCommunicationCalling/CallingSDKEventsHandler.swift:231]: Fixed

**Comment by JoshuaLai** [comment]
Chatted offline about this to understand the design a bit more. I think my suggestion is to instead of following the CallSDK state with the 
```swift 
enum CallDiagnosticQuality: Int {
    case unknown
    case good
    case poor
    case bad
}
```
We map it to something more specific to the UI composite. Just food for thought

**Comment by lucianopa-msft** [comment]
Closing to open the PR from the correct branch

**Pull request closed without merging**